### PR TITLE
[Merged by Bors] - allow specifying pvc parameters

### LINKED ISSUE #67: Make resources configurable (CPU, memory, PVCs)
Especially StorageClass and size

This is done when
- [ ] resources are configurable according to the standards set by operator-rs
- [ ] documentation has been added
- [ ] integration test has been added

### PR DESCRIPTION
**Pull request by razvan**
## Description

Fixes #67 

## Review Checklist
- [x] Code contains useful comments
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by nightkr** [comment]
@soenkeliebau has been working on integrating something similar into his resource management scheme (over at https://github.com/stackabletech/operator-rs/compare/feat/resource_config2, don't think he made a proper PR from what I can tell?). Might be worth talking to him about aligning this.

**Comment by razvan** [comment]
> @soenkeliebau has been working on integrating something similar into his resource management scheme (over at https://github.com/stackabletech/operator-rs/compare/feat/resource_config2, don't think he made a proper PR from what I can tell?). Might be worth talking to him about aligning this.

Didn't know that. Thanks for the pointer.

File: CHANGELOG.md
Comment (maltesander) [inline CHANGELOG.md]:
```suggestion
- PVCs for data storage, cpu and memory limits are now configurable ([#164]).
```

File: Cargo.toml
Comment (maltesander) [inline Cargo.toml]: remove?

File: docs/modules/ROOT/pages/usage.adoc
Comment (maltesander) [inline docs/modules/ROOT/pages/usage.adoc]:
```suggestion
You can mount volumes where data is stored by specifiying https://kubernetes.io/docs/concepts/storage/persistent-volumes[PersistentVolumeClaims] for each individual role group:
```

File: rust/crd/Cargo.toml
Comment (maltesander) [inline rust/crd/Cargo.toml]: Just a reminder to switch back on the tag (i guess 0.19.0?).

File: rust/crd/src/lib.rs
Comment (maltesander) [inline rust/crd/src/lib.rs]: I dont get the benefit of that default? The only difference to deriving is pointing to a config map that does not exist? Whats the difference to the zookeeper_config_map_name defaulting to "" ?

File: rust/crd/src/lib.rs
Comment (maltesander) [inline rust/crd/src/lib.rs:268]: Could we get more readable variable names here?

File: rust/crd/src/lib.rs
Comment (maltesander) [inline rust/crd/src/lib.rs:270]: I would not make this a vec just to fit the `volume_claim_templates`. Decreases testability etc.

File: rust/operator/src/hdfs_controller.rs
Comment (maltesander) [inline rust/operator/src/hdfs_controller.rs]: Can we change the variable `rr` to `resources` or `resource_requirements`?

File: rust/crd/src/lib.rs
Comment (maltesander) [inline rust/crd/src/lib.rs]: Is this for a linter or copy paste?
```suggestion
        replicas: 1";
```

File: rust/crd/src/lib.rs
Comment (maltesander) [inline rust/crd/src/lib.rs]: remove `_rolegroup_ref`?

File: rust/crd/src/lib.rs
Comment (maltesander) [inline rust/crd/src/lib.rs]: remove?

**Review by maltesander — COMMENTED** [review]
Just some comments. Ill test tomorrow. Can we have a readme for the kuttl tests? Its not explained how to properly run that with the test definition yaml?

File: rust/operator/src/hdfs_controller.rs
Comment (razvan) [reply rust/operator/src/hdfs_controller.rs]: Done.

File: rust/crd/src/lib.rs
Comment (razvan) [reply rust/crd/src/lib.rs]: Done.

File: rust/crd/src/lib.rs
Comment (razvan) [reply rust/crd/src/lib.rs]: Done.

File: rust/crd/src/lib.rs
Comment (razvan) [reply rust/crd/src/lib.rs:270]: I don't see how this decreases testability.

File: rust/crd/src/lib.rs
Comment (razvan) [reply rust/crd/src/lib.rs]: It was a leftover.

File: rust/crd/Cargo.toml
Comment (razvan) [reply rust/crd/Cargo.toml]: After that branch is merged and a new release is done.

File: rust/crd/src/lib.rs
Comment (maltesander) [reply rust/crd/src/lib.rs:270]: Ok let me rephrase then. Will there be more than one element in that vector in the future? If not, i dont see the point in returning a vec (simply to save on the `vec![xxx]`) when building the template?

File: Cargo.toml
Comment (razvan) [reply Cargo.toml]: done.

File: rust/crd/src/lib.rs
Comment (razvan) [reply rust/crd/src/lib.rs:270]: HDFS supports multiple volumes for it's data folders, so we might support that in the future. But to me, this is about encapsulation: I have one clear interface (this function) that returns PVCs and don't have to fiddle with them later in the code to pass them on to K8S.

Comment (maltesander) [reply rust/crd/src/lib.rs:270]: Ok did not know that. Then i resolve this.

**Review by maltesander — COMMENTED** [review]
Tested and works. @soenkeliebau should approve once he had a chance to look over this.
Locally the tests failed the first time (timeout 5min) and worked after the images were downloaded. Maybe we can increase the timeout for the install step?

**Review by soenkeliebau — APPROVED** [review]
I have looked it over and tested it, looks good to me.
One thing I noticed is that  the tests specify resources but don't check if they get set in the pods / pod templates correctly. That migth be an improvement down the line that we could add.
But I am also happy for this to go in as is.

**Comment by razvan** [comment]
bors merge

**Pull request closed without merging**
